File: dart/dart-sample/sample/oop/class.dart
// ignore_for_file: unused_local_variable

void main(List<String> args) {
  for (int i = 0; i < 5; i++) {
    Person('no. $i');
  }
  print(Person.getPeopleCount());
}

class Person {
  String name;
  // class variables
  static int peopleCount = 0;
  Person(this.name) {
    peopleCount++;
  }
  static getPeopleCount() => peopleCount;
}

File: nodejs/typescript-sample/oop/class.ts
class Person {
  // private fields
  #name: string
  #age: number

  static #peopleCount = 0

  // constructors
  constructor(name: string, age: number) {
    this.#age = age
    this.#name = name
    Person.#peopleCount++
  }

  // getters and setters
  get name (): string {
    return this.#name
  }
  set name (value: string) {
    this.#name = value
  }

  get age (): number {
    return this.#age
  }
  set age (value: number) {
    this.#age = value
  }

  // methods
  sleep (): void {
    console.log(`have a good night, ${this.#name}`)
  }
  static getPeopleCount (): number {
    return Person.#peopleCount
  }
}

const p = new Person('zs', 18)
p.name = 'john'

p.sleep()
new Person('jj', 6)
console.log(Person.getPeopleCount())
